### ISSUE TITLE
some corrections on lists that (some break existing)

### ISSUE THREAD

**Issue by og76** [JacquesLucke/animation_nodes#452]
as per #451
 _Should this be on a separate branch? Along with #450 ?(also modified here)
Cause I think some break older files by touchin the sockets._

mainly about 
 - tuple / limited list case (not working or even throwin exception)  
 - (ui) adding *advanced invokeSocketTypeChooser* (where not present yet)
 - (ui) auto type to integer for limitedList when plug a sock (where not present yet)

the nodes in red used to give errors on tuple, the pink used to return tuple (in an integer socket)
![list correction1](https://cloud.githubusercontent.com/assets/11709974/14068621/be22c0fa-f491-11e5-8b98-e2e21588ab47.PNG)

nodes:

- append (may break?) 
 - renamed sockets to inList, (does this break old?)
 - now accept tuple (not workin before, **error**) !
 - (ui) auto change to integer for limited type

- get random
 - (ui) advanced invokeSocketTypeChooser
 - (ui) auto change to integer for limited type

- remove
 - now accept tuple (**error ** before, for del etc)
 - (ui) advanced invokeSocketTypeChooser
 - (ui) auto change to integer for limited type

- search
 - (ui) auto change to integer for limited type

- shuffle (may break?)
 - renamed sockets to inList, does this break old?
 - now accept tuple (**error ** before)
 - (ui) advanced invokeSocketTypeChooser
 - (ui) auto change to integer for limited type

- reverse 
 - (ui) advanced invokeSocketTypeChooser
 - (ui) auto change to integer for limited type

- slice (may break?) + should I leave it out and change in #450 ? or merge and change after?
 - renamed sockets to inList, does this break old?
 - now accept tuple (before **was returning a tuple**! `(1,2,3)[:1] = (1,)` )
 - changed all type stuff to ListDataType, no baseDataType needed, no element socket (simplified)
 - (ui) auto change to integer for limited type

- shift (as above, but not renamin socket, so maybe not break?)
 - now accept tuple (before **was returning a tuple**! `(1,2,3)[1:] + (1,2,3)[:-1] = (3,1,2)` )
 - (ui) advanced invokeSocketTypeChooser
 - (ui) auto change to integer for limited type

- boolean list:
 - (ui) advanced invokeSocketTypeChooser
 - (ui) auto change for limited type

-

have tested for loops also, to see if the append/set/remove or so still work on the original input (dataIsModified?) and seams ok. 
The fact that poly/edge index becomes a list (not tuple) is not an issue for mesh data and further nodes if the case. 
![list correction2](https://cloud.githubusercontent.com/assets/11709974/14068633/2b13d8ac-f492-11e5-8d51-f08b5a687e8c.PNG)

ps:
I'd prefer to not push them one by one, but if u find it hard to follow we'll see

**Comment by JacquesLucke** [comment]
hmm looks like the better solution would be to implement a Convert to Integer List node... this would solve all these problems without breaking nodes + we get a (little) better performance.
so we should remove the concept of limited lists and just use converter nodes..

**Comment by og76** [comment]
or maybe we should not use tuples at all. 
lists of lists work very well for the mesh data poly/edge, unless they add a bit when reading from blender

1. err, what better performance with an extra node?
and how is it not gonna break if we change the socket files / socket info ?
(truth is, I don't like extra nodes, and converts are like that, but that is another story)

2. this is also about havin advanced assign type, or updatin some other parts


3. ok, should all these be targeted and discussed in another branch? 
in ex also the pivot or others.
updatin the master bit by bit is not a good idea, break or not.

**Comment by JacquesLucke** [comment]
I would like to keep using tuples for edge/poygon indices because:
1. requires less space than lists
2. prevents users from creating invalid mesh data. (can easily crash blender...)

the getExecutionCode concept is so powerful because it removes all overhead. when we rename the sockets to inList and outList an extra assignment is needed each time certain nodes are called. (at least I think so)
I know, converter nodes are not very nice but this here is something people rarely do.. I still think a converter is the best choice.

I think I'm able to tell which changes will break and which not. and I think adding this converter and removing the concept of limited lists won't break existing files.

I'm underwayy now till Wednesday.

**Comment by og76** [comment]
ok. could be.
just as a note, I have not changed any inList/outList as i/o pair, where inList/inList was goin for both i/o.
So I kept the current state of matching i/o name (or var) including dataIsModified etc.
just renamed inlist where the name was list (as the function). 

hm. actually, name as function or prop or type,  should never be ..

even without break, I still think we should work on a different v1.5a or something, that would be "released" one time with all minor changes. the very way blender does for correcting necessary bugs
This will help people follow the version. 
Even for new nodes. Instead of saying, hey! there are some new stuff, go dld, and tomorrow yet another, go dld ...

**Comment by og76** [comment]
ps. it was a great victory anyway, I managed to push this directly from git desktop :D

**Comment by JacquesLucke** [comment]
I removed the concept of limited lists and implemented the Convert to Integer List node. It can convert edge/polygon indices and float list.

Also I renamed the socket in the Set List Element node again.
I decided to keep "list" as identifier as this is a very local local variable. It is replaced by another name before it is copied into the execution code. It only limits us in the way that we cannot use the list function inside of this node but I think that is okay.

1bc9cfce35c80cb43d87681ecb65b2d2c9f5ca22

I agree that we should do a v1.5.1 release within the next three weeks with some bugfixes and minor additions. I created a branch for that.
